File: 0784-letter-case-permutation/0784-letter-case-permutation.java
class Solution {
    
    List<String> ans = new ArrayList<>();
    
    public void recurs(String s, int idx, int n) {
        if(idx == n) {
            if(!ans.contains(s)) {
                ans.add(s);
            }
            return;
        }
        
        if(Character.isLetter(s.charAt(idx))){
            recurs(s.substring(0, idx)+Character.toLowerCase(s.charAt(idx))+s.substring(idx+1, n), idx+1, n);
            recurs(s.substring(0, idx)+Character.toUpperCase(s.charAt(idx))+s.substring(idx+1, n), idx+1, n);
        } else {
            recurs(s, idx+1, n);
        }
        
    }
    public List<String> letterCasePermutation(String s) {
        recurs(s, 0, s.length());
        return ans;
    }
}

File: 0784-letter-case-permutation/0784-letter-case-permutation.py
class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        
        ans = set()
        
        def recur(s, idx, n):
            if idx == n:
                ans.add(s)
                return
            
            if s[idx].isalpha():
                recur(s[:idx]+s[idx].lower()+s[idx+1:], idx+1, n)
                recur(s[:idx]+s[idx].upper()+s[idx+1:], idx+1, n)
            else:
                recur(s, idx+1, n)
        
        recur(s, 0, len(s))
        return ans